rust-lang/cargo
# refactor: abstract `std::fs` away from on-disk index cache

### What does this PR try to resolve?

Part of https://github.com/rust-lang/cargo/issues/6908

This abstracts `std::fs` so that we can integrate SQLite into on-disk cache in follow-ups.

### How should we test and review this PR?


* Make sure the documentation still render well.
* Shouldn't have any behavior change. Only a bit of extra allocations added.

### Additional information
